src/awkward/_connect/jax/trees.py

```diff
@@ -6,162 +6,114 @@
-        #     nplike=Numpy.instance(),
-        # )
-
-        return buffers, AuxData(
-            layout=layout,
-            is_highlevel=is_highlevel,
-            behavior=behavior_of(obj),
+        # decompose
+        form, length, buffers = ak.operations.to_buffers(layout)
+
+        # we need to split buffers into all "data" buffers and all others (e.g. index, offsets, etc.)
+        data_buffers, other_buffers = split_buffers(buffers)
+
+        # now we need to flatten the data buffers
+        data_buffer_keys, data_flat_buffers = zip(*data_buffers.items())
+        return data_flat_buffers, AuxData(
+            data_buffer_keys=data_buffer_keys,
+            other_buffers=other_buffers,
+            form=form,
+            length=length,
+            ctx=ctx,
+            highlevel=not isinstance(obj, Content),
         )
 
-    @property
-    def layout(self) -> contents.Content | record.Record:
-        return self._layout
-
-    @property
-    def behavior(self) -> dict | None:
-        return self._behavior
-
-    @property
-    def is_highlevel(self) -> bool:
-        return self._is_highlevel
-
-    def unflatten(self, buffers: tuple) -> T:
-        for buffer in buffers:
-            # Check that JAX isn't trying to give us float0 types
-            dtype = getattr(buffer, "dtype", None)
-            if dtype == np.dtype([("float0", "V")]):
-                raise TypeError(
-                    f"a buffer with the dtype {buffer.dtype} was encountered during unflattening. "
-                    "JAX uses this dtype for the tangents of integer/boolean outputs; these cannot "
-                    "reasonably be differentiated. Make sure that you are not computing the derivative "
-                    "of a boolean/integer (array) valued function."
-                )
-
-        # Replace the mixed NumPy-JAX layout leaves with the given buffers (and use the JAX nplike)
-        layout = replace_all_buffers(
-            self._layout, list(buffers), backend=JaxBackend.instance()
+    def unflatten(self, data_buffers: tuple) -> T:
+        # reconstitute data buffers
+        data_buffers = dict(zip(self._data_buffer_keys, data_buffers))
+
+        # combine data buffers with other buffers
+        buffers = {**self._other_buffers, **data_buffers}
+
+        # reconstruct layout
+        layout = ak.operations.from_buffers(
+            self._form, self._length, buffers, backend="jax"
         )
-        return wrap_layout(
-            layout, behavior=self._behavior, highlevel=self._is_highlevel
+
+        # wrap layout
+        return self._ctx.wrap(
+            layout,
+            highlevel=self._highlevel,
+            allow_other=True,
         )
 
     def __eq__(self, other: AuxData) -> bool:
-        return self.layout.is_equal_to(
-            other.layout, index_dtype=False, numpyarray=False
-        )
+        # should this test for more?
+        return self._form == other._form and self._length == other._length
```

Comment:
Perhaps use `form.is_equal_to`, and pass `all_parameters=True` and `form_key=True`?

Reply:
good point! :+1: